### PR TITLE
CSHARP-1326: Fix edge case in Implements and TryGetIEnumerableGenericInterface.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Misc/TypeExtensions.cs [hunk 1]
```diff
@@ -32,10 +32,14 @@ public static Type GetIEnumerableGenericInterface(this Type enumerableType)
...
         {
-            Type interfaceDefinition = null;
-            if (@interface.IsGenericType())
+            if (type == @interface)
             {
-                interfaceDefinition = @interface.GetGenericTypeDefinition();
+                return true;
+            }
+
+            if (type.IsGenericType() && type.GetGenericTypeDefinition() == @interface)
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Misc/TypeExtensions.cs:40]: For example, we want `typeof(IEnumerable<T>).Implements(typeof(IEnumerable<>))` to return `true`.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Misc/TypeExtensions.cs [hunk 2]
```diff
@@ -32,10 +32,14 @@ public static Type GetIEnumerableGenericInterface(this Type enumerableType)
 
         public static bool Implements(this Type type, Type @interface)
         {
-            Type interfaceDefinition = null;
-            if (@interface.IsGenericType())
+            if (type == @interface)
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Misc/TypeExtensions.cs:35]: We want `typeof(ISomething>).Implements(typeof(ISomething))` to return `true`.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Misc/TypeExtensions.cs [hunk 3]
```diff
@@ -104,6 +108,12 @@ public static bool TryGetIDictionaryGenericInterface(this Type type, out Type id
 
         public static bool TryGetIEnumerableGenericInterface(this Type type, out Type ienumerableGenericInterface)
         {
+            if (type.IsGenericType() && type.GetGenericTypeDefinition() == typeof(IEnumerable<>))
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Misc/TypeExtensions.cs:111]: We want `typeof(IEnumerable<T>>).TryGetEnumerableGenericInterface(out var ienumerableGenericInterface)` to return `true` and set the out parameter to `typeof(IEnumerable<T>))`.

**Review by rstam — COMMENTED** [review]
A couple of edge cases I missed when implementing CSHARP-1326 resulted in a failed test.

**Review by JamesKovacs — APPROVED** [review]
LGTM
